[core] add support for integration span hooks

brettlangdon opened the pull request:
This PR adds the initial prototype for registering and emitting hooks to be called that allow users to customize spans created by integrations.

This also includes changes to emit the `falcon.request` span from our `falcon` integration as an example.


```python
from ddtrace import config

@config.falcon.hooks.on('request')
def on_falcon_request(span, request, response):
    span.set_tag('custom', 'tag')
```

brettlangdon commented on ddtrace/contrib/falcon/middleware.py:
```diff
@@ -76,6 +76,12 @@ def process_response(self, req, resp, resource, req_succeeded=None):
+
+        # Emit span hook for this response
+        # DEV: Emit before closing so they can overwrite `span.resource` if they want
+        self.tracer.emit(span, req, resp)
```
@labbati wondering if this is the best place for this or not?

palazzem commented on ddtrace/tracer.py:
```diff
@@ -452,3 +458,70 @@ def set_tags(self, tags):
+            return func
+        return wrapper
+
+    def emit(self, span, *args, **kwargs):
```
I guess this API is used only by our instrumentation right? I don't see the benefit of using it from user code (or maybe yes?). My question is more about if we want to make it "public" or if we want to keep these primitives internal so that only `tracer.on()` is explained and documented.

brettlangdon replied:
Yeah, I can't really imagine why a user would use this.

I'll move to `def _emit()`, can always make it public in the future if we need to, easier than moving it to being internal.

palazzem commented on ddtrace/tracer.py:
```diff
@@ -452,3 +458,70 @@ def set_tags(self, tags):
+        :raises: TypeError
+        """
+        if not isinstance(span, Span):
+            raise TypeError('Tracer.emit expected a `ddtrace.span.Span` instead got "{}"'.format(span))
```
In general we should never fail so exceptions must never bubble up to users' code. Because this is used by our instrumentation (and we cover it with unittests / integration tests), it's fine.

palazzem commented on ddtrace/tracer.py:
```diff
@@ -452,3 +458,70 @@ def set_tags(self, tags):
+            try:
+                func(span, *args, **kwargs)
+            except Exception as err:
+                log.error('error while calling "{}" hook function "{}": {}'.format(span.name, fname, err))
```
We don't have a good way to throttle logging. If we use `log.error` (that is semantically correct) we risk to write 1k log entries per second if we hit a bug and this is in users' code critical path.

I think we should keep it as `log.debug` until we throttle our logging (i.e. not more than 5 log entries in 10 seconds).

brettlangdon commented:
I updated here based on feedback.

- `Tracer.emit` -> `Tracer._emit`
- Do not raise any exceptions
- Use `log.debug` instead of `log.error`

brettlangdon commented:
Removed WIP tag, since this passed general API review from the team, this is good to be reviewed/merged.

Kyle-Verhoog commented on ddtrace/tracer.py:
```diff
@@ -452,3 +458,70 @@ def set_tags(self, tags):
+        # Iterate and call all hooks for this span name
+        for func in self._hooks[span.name]:
+            fname = func_name(func)
+            log.debug('calling "{}" hook function "{}"'.format(span.name, fname))
```
Maybe this should be a `log.info`? Seems like it would be called a lot in some cases.

brettlangdon replied:
probably wouldn't make a difference with logging, since most of the time we do `DATADOG_TRACE_DEBUG=true` which sets the log level to `debug`, so it would be traced no matter what.

also, if the user sets up default logging `logging.baseConfig(level=logging.INFO)` when these will start to show up for their code.

Kyle-Verhoog commented on tests/test_tracer.py:
```diff
@@ -485,6 +485,80 @@ def test_start_child_from_context():
     eq_(child, child._context._current_span)
 
 
+def test_tracer_hook():
```
We should probably have some test coverage for passing args/kwargs with `_emit`.l

brettlangdon commented:
@rochdev had a great suggestion that we also (optionally) allow a user to scope their hook to a specific service.

For example:

```python
@tracer.on('falcon.request', service='my-custom-service'):
def on_falcon_request(span, req, res):
    pass
```

Where this hook only gets called if the `falcon.request` span has a service of `my-custom-service`.

The default will always be `service=None`, which will always get run for any span with the name `falcon.request`.

Kyle-Verhoog commented:
This is a great idea.

brettlangdon commented:
I updated this code for the changes that @rochdev and I were discussing.

The main difference here is that we are now going to register the hooks as part of configuration instead of directly on the tracer.

The reason for this is to ensure we don't have hook names that will clash between integrations.

As well, moving this way will help once we move to allow instance specific configuration as well.


I copied some of @labbati's changes from his HTTP header PR for `IntegrationConfig`. It is brought here as an example, I want to break it out into it's own PR, since we now have 2 PRs that are dependent on it. I also added in `AttrDict`.


New version:

```python
from ddtrace import config

@config.falcon.hooks.request
def on_falcon_request(span, request, response):
    pass


config.falcon._emit('request', span, request, response)
```

labbati commented on ddtrace/settings.py:
```diff
@@ -102,8 +104,8 @@ def __init__(self, global_config, *args, **kwargs):
+        self.hooks = Hooks()
 
     def __deepcopy__(self, memodict=None):
         new = IntegrationConfig(self.global_config, deepcopy(dict(self)))
```
Should we handle the deep copy of hooks as well?

brettlangdon replied:
oh yes, we need to, great catch!

brettlangdon replied:
```suggestion
        new = IntegrationConfig(self.global_config, deepcopy(dict(self)))
        new.hooks = self.hooks
```

labbati commented on ddtrace/settings.py:
```diff
@@ -115,5 +117,133 @@ def __repr__(self):
+    Example::
+
+        @config.falcon.hooks.request
+        def on_reqiest(span, request, resposne):
```
`on_request`

brettlangdon replied:
```suggestion
        def on_request(span, request, response):
```

labbati commented on ddtrace/settings.py:
```diff
@@ -115,5 +117,133 @@ def __repr__(self):
+            try:
+                func(span, *args, **kwargs)
+            except Exception as e:
+                log.debug('Failed to run hook {} function {}: {}'.format(hook, func, e))
```
Shouldn't it be `log.error`?

brettlangdon replied:
I'll add a comment but no, basically because we use the built in logger we don't have a way to throttle any logging we do.

```suggestion
                # DEV: Use log.debug instead of log.error until we have a throttled logger
                log.debug('Failed to run hook {} function {}: {}'.format(hook, func, e))
```

Kyle-Verhoog commented on ddtrace/settings.py:
```diff
@@ -115,5 +118,134 @@ def __repr__(self):
+        :returns: A function decorator used to register a hook
+        :rtype: function
+        """
+        return self.register(key)
```
There is some ergonomic benefit added here. However I think this makes things less clear in the long run, especially if we begin adding other methods on `Hooks`. (What if we want to support a hook for something called  `'deregister'`?)

brettlangdon commented:
@Kyle-Verhoog I have updated the register API as you suggested
